### PR TITLE
Adds debugging view to help inspect result logic (closes #125).

### LINKED ISSUE #125: Create a debugging endpoint to help understand
`/debug/?q=f` should show a nice view of all the data we gathered to inform our recommendation-making process.

### PR DESCRIPTION
**Pull request by chuckharmston**
Video of the output: https://chchc.ch/0y3XY8

Was writing similar code to help get to the bottom of some searches @nchapman helped me look into, figured I'd go the extra mile to make it easily accessible, even in production.

No need to review or merge now, but this is ready. r? @6a68

### CODE REVIEW

File: recommendation/search/recommendation.py [hunk 1]
```diff
@@ -16,8 +16,12 @@ class SearchRecommendation(object):
       sequentially determine if they match the result, then enhance the result
       if so.
     """
-    def __init__(self, query):
+    def __init__(self, query, debug=False):
+        self.debug = debug
         self.query = query
+        self.recommendation = None
+        self.suggestion_engine = self.get_suggestion_engine()
+        self.query_engine = self.get_query_engine()
```
Comment (hannosch) [inline recommendation/search/recommendation.py]: If you are by-passing the getter methods anyways, maybe you could just make the two engines class level attributes and avoid the getter altogether.

File: recommendation/search/query/yahoo.py [hunk 2]
```diff
@@ -33,7 +33,10 @@ def _oauth_url(self, query):
 
     @memorize(prefix='yahoo')
     def fetch(self, query):
-        return requests.get(self._oauth_url(query)).json()
+        response = requests.get(self._oauth_url(query))
+        print(response.status_code)
+        print(response.json())
```
Comment (hannosch) [inline recommendation/search/query/yahoo.py]: These print statements don't look they belong in the final code.

File: recommendation/search/recommendation.py (hunk 1 above)
Comment (chuckharmston) [inline recommendation/search/recommendation.py]: The goal is to eventually support negotiation in the getters; i.e. have the add-on send the user's default search provider, and respect that even on the server side. We aren't quite there yet, but the hooks are :)

File: recommendation/search/query/yahoo.py (hunk 2 above)
Comment (chuckharmston) [inline recommendation/search/query/yahoo.py]: That is absolutely correct. Good catch!

File: recommendation/search/recommendation.py (hunk 1 above)
Comment (hannosch) [inline recommendation/search/recommendation.py]: Sure, I thought as much, but isn't that what `@property` is for ;)

This just struck me as odd, as you later on changed the code to use `self.suggestion_engine` directly, so it isn't either getter-style nor proprerty-style but both at the same time.

**Comment by 6a68** [comment]
R+, merge at will

**Pull request merged**